source/features/fork-source-link-same-view.tsx

```diff
@@ -0,0 +1,66 @@
+import select from 'select-dom';
+import * as pageDetect from 'github-url-detection';
+
+import features from '.';
+import * as api from '../github-helpers/api';
+import GitHubURL from '../github-helpers/github-url';
+import {getRepoURL} from '../github-helpers';
+
+const checkIfFileExists = async (url: GitHubURL) => {
+	const {repository} = await api.v4(`
+		repository(owner: "${url.user}", name: "${url.repository}") {
+			file: object(expression: "${url.branch}:${url.filePath}") {
+				... on Blob {
+					id
+				}
+			}
+		}
+	`);
+
+	if (!repository.file) {
+		const source = select<HTMLAnchorElement>(`[href$="${url.pathname}" i]`)!;
+		source.pathname = '/' + source.textContent!.trim();
+	}
+};
+
+// eslint-disable-next-line import/prefer-default-export
+export function createHEADLink(baseRepo: string): string {
+	if (pageDetect.isRepoRoot() || !(pageDetect.isSingleFile() || pageDetect.isRepoTree() || // There will be no change to the link
+		baseRepo.toLowerCase() === getRepoURL()) // Your there already
+	) {
+		return '/' + baseRepo;
+	}
+
+	const [user, repository] = baseRepo.split('/');
+	const url = new GitHubURL(location.href).assign({
+		user,
+		repository,
+		branch: 'HEAD'
+	});
+
+	if (pageDetect.isSingleFile()) {
+		// If the file does not exists update the link, but dont await it.
+		void checkIfFileExists(url);
+	}
+
+	return url.pathname;
+}
+
+function init(): void {
+	const forkSource = select<HTMLAnchorElement>('.fork-flag .text a')!;
+	forkSource.pathname = createHEADLink(forkSource.textContent!);
```

Comment:
Don't overthink it. Why set, then check, then search the link again and reset it?

```suggestion
	const forkSource = select<HTMLAnchorElement>('.fork-flag .text a')!;
	const sameViewUrl = createHEADLink(forkSource.textContent!);
	const response = await fetch(sameViewUrl, {method: 'HEAD'});
	if (response.ok) {
		forkSource.pathname = sameViewUrl
	}
```

Reply:
Better yet, skip everything if no change is necessary:

```suggestion
	const forkSource = select<HTMLAnchorElement>('.fork-flag .text a')!;
	const sameViewUrl = createHEADLink(forkSource.textContent!);
	if (sameViewUrl === forkSource.pathname) {
		return false
	}

	const response = await fetch(sameViewUrl, {method: 'HEAD'});
	if (response.ok) {
		forkSource.pathname = sameViewUrl
	}
```

Reply:
Quick question are we only checking only if its a single file?

Reply:
Also can we do a v4 api request for 2 reasons. I think its lighter and avoids the 404 console error. (I hate those 😄 )